routes/runtime/github.rb

```diff
@@ -0,0 +1,128 @@
+        .exclude(committed_at: nil)
+        .where(key: keys, version: version, scope: scopes).all
+        .min_by { keys.index(_1.key) + (scopes.index(_1.scope) * keys.size) }
+
+      fail CloverError.new(204, "NotFound", "No cache entry") if entry.nil?
+
+      entry.update(last_accessed_at: Time.now, last_accessed_by: runner.id)
+      signed_url = repository.url_presigner.presigned_url(:get_object, bucket: repository.bucket_name, key: entry.blob_key, expires_in: 900)
+
+      {
+        scope: entry.scope,
+        cacheKey: entry.key,
+        cacheVersion: entry.version,
+        creationTime: entry.created_at,
+        archiveLocation: signed_url
+      }
+    end
+
+    r.on "caches" do
+      # listCache
+      r.get true do
+        key = r.params["key"]
+        fail CloverError.new(204, "NotFound", "No cache entry") if key.nil?
+
+        scopes = [runner.workflow_job&.dig("head_branch"), repository.default_branch].compact
+        entries = repository.cache_entries_dataset
+          .exclude(committed_at: nil)
+          .where(key: key, scope: scopes).all
+
+        {
+          totalCount: entries.count,
+          artifactCaches: entries.map do
+            {
+              scope: _1.scope,
+              cacheKey: _1.key,
+              cacheVersion: _1.version,
+              creationTime: _1.created_at
+            }
+          end
+        }
+      end
+
+      repository.setup_blob_storage unless repository.access_key
+
+      # reserveCache
+      r.post true do
+        key = r.params["key"]
+        version = r.params["version"]
+        size = r.params["cacheSize"].to_i
+        fail CloverError.new(400, "InvalidRequest", "Wrong parameters") if key.nil? || version.nil? || size == 0
+
+        unless (scope = runner.workflow_job&.dig("head_branch"))
+          # YYYY: If the webhook not delivered yet, we can try to get the branch from the API
+          Clog.emit("The runner does not have a workflow job") { {no_workflow_job: {ubid: runner.ubid, repository_ubid: repository.ubid}} }
+          fail CloverError.new(400, "InvalidRequest", "No workflow job data available")
+        end
+
+        entry = GithubCacheEntry.create_with_id(repository_id: runner.repository.id, key: key, version: version, size: size, scope: scope, created_by: runner.id)
+
+        upload_id = repository.blob_storage_client.create_multipart_upload(bucket: repository.bucket_name, key: entry.blob_key).upload_id
+        entry.update(upload_id: upload_id)
+
+        max_chunk_size = 32 * 1024 * 1024 # 32MB
+        presigned_urls = (1..size.fdiv(max_chunk_size).ceil).map do
+          repository.url_presigner.presigned_url(:upload_part, bucket: repository.bucket_name, key: entry.blob_key, upload_id: upload_id, part_number: _1, expires_in: 900)
+        end
+
+        {
+          uploadId: upload_id,
+          presignedUrls: presigned_urls,
+          chunkSize: max_chunk_size
+        }
+      end
+
+      # commitCache
+      r.post "commit" do
```

Comment:
What happens if chunks can be uploaded partially? Does the client send a request to separate endpoint? How do we make sure that there is no dangling chunk.

Reply:
The client uploads all chunks to the blob storage, each upload request returning an etag as a response. The client then sends these etags for each chunk as a list to the backend in the commit cache request. Finally, the backend sends a complete multipart request to the blob storage with these etags. If any etag is incorrect or missing, blob storage will return failure for complete multipart request.

Reply:
My question was more about dangling chunks in case of partial upload. Your explanation offline (aborting after 7 days) answers my question.